File: bitDP.py
#submit https://onlinejudge.u-aizu.ac.jp/solutions/problem/DPL_2_A/review/9037879/T21140031/PyPy3
class Edge:
    def __init__(self,to,cost):
        self.to = to
        self.cost = cost

INF = 1 << 30

n, m = map(int,input().split()) #n頂点m辺
g = [[] for _ in range(n)] #隣接リスト
for _ in range(m):
    s,t,d = map(int,input().split())
    g[s].append(Edge(t,d))

dp = [[INF] * n for _ in range(1 << n)]

for e in g[0]:
    dp[1 << e.to][e.to] = e.cost

for k in range(n):#k:立っているbit数
    for i in range(1 << n):
        if i.bit_count() != k: continue #立っているbit数がkでなかった
        for j in range(n):
            if i >> j & 1 == 0: continue #訪れた頂点の集合(i)に頂点jが含まれていなかった
            if dp[i][j] == INF: continue
            
            for e in g[j]:
                if i >> e.to & 1 == 1: continue #行き先がすでに到達済みだった
                #遷移
                dp[i + (1 << e.to)][e.to] = min(dp[i + (1 << e.to)][e.to],dp[i][j] + e.cost)

if dp[(1 << n)-1][0] == INF:
    print(-1)
else:
    print(dp[(1 << n)-1][0])

File: 20240331_bit全探索_bitDP/提出コード/bitDP.cpp
//submit  https://onlinejudge.u-aizu.ac.jp/solutions/problem/DPL_2_A/review/9037754/T21140031/C++17
#include <bits/stdc++.h>
using namespace std;

const int INF = 1 << 30;

int main(){
  ios::sync_with_stdio(false);cin.tie(nullptr);
  int n,m;cin >> n >> m;// n頂点m辺
  vector<vector<pair<int,int>>> g(n);
  for(int i = 0;i < m;i++){
    int s,t,d;cin >> s >> t >> d;
    g[s].push_back({t,d});
  }

  vector<vector<int>> dp(1 << n,vector<int>(n,INF));
  //初期化
  for(const auto &[to,cost] :g[0]){
    dp[1 << to][to] = cost;
  }

  for(int k = 0;k < n;k++){//k:立っているbit数
    for(int i = 0;i < (1 << n);i++){
      if(__builtin_popcount(i) != k)continue;//立っているbit数がkでなかった
      for(int j = 0;j < n;j++){
        if((i >> j & 1) == 0)continue;//訪れた頂点の集合(i)に頂点jが含まれていなかった
        if(dp[i][j] == INF)continue;

        for(const auto &[to,cost] :g[j]){
          if((i >> to & 1) == 1)continue;//行き先がすでに到達済みだった
          //遷移
          dp[i + (1 << to)][to] = min(dp[i + (1 << to)][to],dp[i][j] + cost);
        }
      }
    }
  }

  if(dp[(1 << n) -1][0] == INF){
    cout << -1 << endl;
  }else{
    cout << dp[(1 << n) -1][0] << endl;
  }
}